Add support for using foreign type units in .debug_names.

clayborg opened the pull request:
This patch adds support for the new foreign type unit support in .debug_names. Features include:
- don't manually index foreign TUs if we have info for them
- only use the type unit entries that match the .dwo files when we have a .dwp file
- fix crashers that happen due to PeekDIEName() using wrong offsets

clayborg committed:
Add support for using foreign type units in .debug_names.

This patch adds support for the new foreign type unit support in .debug_names. Features include:
- don't manually index foreign TUs if we have info for them
- only use the type unit entries that match the .dwo files when we have a .dwp file
- fix crashers that happen due to PeekDIEName() using wrong offsets

bulbazord commented on lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp:
```diff
@@ -273,6 +301,44 @@ void DebugNamesDWARFIndex::GetFullyQualifiedType(
+              cu_die.GetAttributeValueAsString(DW_AT_dwo_name, nullptr);
+          llvm::StringRef tu_dwo_name =
+              tu_die.GetAttributeValueAsString(DW_AT_dwo_name, nullptr);
+          if (cu_dwo_name != tu_dwo_name)
```
Is it possible for `cu_dwo_name` and `tu_dwo_name` to both be empty/invalid?

bulbazord commented on lldb/source/Plugins/SymbolFile/DWARF/ManualDWARFIndex.h:
```diff
@@ -21,9 +21,11 @@ class SymbolFileDWARFDwo;
       : DWARFIndex(module), m_dwarf(&dwarf),
-        m_units_to_avoid(std::move(units_to_avoid)) {}
+        m_units_to_avoid(std::move(units_to_avoid)),
+        m_type_sigs_to_avoid(type_sigs_to_avoid) {}
```
nit: If we are going to std::move `units_to_avoid`, we should do the same for `m_type_sigs_to_avoid`. It's the same situation.

bulbazord commented on lldb/source/Plugins/SymbolFile/DWARF/SymbolFileDWARF.cpp:
```diff
@@ -693,7 +693,6 @@ llvm::DWARFDebugAbbrev *SymbolFileDWARF::DebugAbbrev() {
   if (debug_abbrev_data.GetByteSize() == 0)
     return nullptr;
 
-  ElapsedTime elapsed(m_parse_time);
```
Why remove this `ElapsedTime`?

bulbazord commented on lldb/source/Plugins/SymbolFile/DWARF/SymbolFileDWARF.cpp:
```diff
@@ -1726,44 +1725,59 @@ lldb::ModuleSP SymbolFileDWARF::GetExternalModule(ConstString name) {
+  // If we are currently in a .dwo file and our file index doesn't match we need
+  // to let the base symbol file handle this.
+  SymbolFileDWARFDwo *dwo = llvm::dyn_cast_or_null<SymbolFileDWARFDwo>(this);
+  if (dwo)
```
nit: Merge these two lines

bulbazord commented on lldb/source/Plugins/SymbolFile/DWARF/SymbolFileDWARF.cpp:
```diff
@@ -1726,44 +1725,59 @@ lldb::ModuleSP SymbolFileDWARF::GetExternalModule(ConstString name) {
-      return DWARFDIE();
-    }
+    SymbolFileDWARFDebugMap *debug_map = GetDebugMapSymfile();
+    if (debug_map) {
```
nit: Merge these

bulbazord commented on lldb/source/Plugins/SymbolFile/DWARF/SymbolFileDWARF.cpp:
```diff
@@ -2717,7 +2731,6 @@ void SymbolFileDWARF::FindTypes(const TypeQuery &query, TypeResults &results) {
       die_context = die.GetDeclContext();
     else
       die_context = die.GetTypeLookupContext();
-    assert(!die_context.empty());
```
Why remove the assertion? Not saying we should keep it, but why doesn't this hold?

bulbazord commented on lldb/test/Shell/SymbolFile/DWARF/x86/dwp-foreign-type-units.cpp:
```diff
@@ -0,0 +1,91 @@
+// RUN: %lldb \
+// RUN:   -o "type lookup IntegerType" \
+// RUN:   -o "type lookup FloatType" \
+// RUN:   -o "type lookup IntegerType" \
```
Why look up IntegerType twice?

dwblaikie commented on lldb/source/Plugins/SymbolFile/DWARF/DWARFDebugInfo.h:
```diff
@@ -58,6 +58,8 @@ class DWARFDebugInfo {
 
   const DWARFDebugAranges &GetCompileUnitAranges();
 
+  const std::shared_ptr<SymbolFileDWARFDwo> GetDwpSymbolFile();
```
Remove const from by-value return. (it messes with move semantics and some other things) - or was this meant to return by reference? - yeah, I guess this latter, the underlying `m_dwarf.GetDwpSymbolFile()` seems to return by const reference, so this function probably should too?

dwblaikie commented on lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp:
```diff
@@ -48,15 +60,31 @@ DebugNamesDWARFIndex::GetUnits(const DebugNames &debug_names) {
+DebugNamesDWARFIndex::GetForeignTypeUnit(const DebugNames::Entry &entry) const {
+  std::optional<uint64_t> type_sig = entry.getForeignTUTypeSignature();
+  if (type_sig)
+    if (auto dwp_sp = m_debug_info.GetDwpSymbolFile())
```
this `auto` should probably be const ref, to avoid inc/dec on the ref counted smart pointer's ref count

dwblaikie commented on lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp:
```diff
@@ -273,6 +301,44 @@ void DebugNamesDWARFIndex::GetFullyQualifiedType(
+        continue;
+      // In order to determine if the type unit that ended up in a .dwp file
+      // is valid, we need to grab the type unit and check the attribute on the
+      // type unit matches the .dwo file. For this to happen we rely on each
+      // .dwo file having its own .debug_names table with a single compile unit
+      // and multiple type units. This is the only way we can tell if a type
+      // unit came from a specific .dwo file.
```
Sounds like this wouldn't work for a merged `.debug_names` table? Could you leave a FIXME/do you plan to fix this?
Oh, it also wouldn't work for any kind of LTO which could have multiple CUs in a single object file/dwo file.

(FYI @cmtice )

dwblaikie replied:
I think our conclusion from previous discussions was to put `DW_IDX_compile_unit`, in addition to the `DW_IDX_type_unit` on the entries in the .debug_names table - and add the CU column to the .debug_tu_index in the dwp file, to match these things up?

ayermolo replied:
> I think our conclusion from previous discussions was to put `DW_IDX_compile_unit`, in addition to the `DW_IDX_type_unit` on the entries in the .debug_names table - and add the CU column to the .debug_tu_index in the dwp file, to match these things up?

We went with adding comp_dir/name to type unit unit die:
https://discourse.llvm.org/t/debuginfo-dwarfv5-lld-debug-names-with-fdebug-type-sections/73445/29?u=ayermolo
A bit of a shortcut so we don't have to deal with non-standar dwp.

ayermolo replied:
> Sounds like this wouldn't work for a merged `.debug_names` table? Could you leave a FIXME/do you plan to fix this? Oh, it also wouldn't work for any kind of LTO which could have multiple CUs in a single object file/dwo file.
> 
> (FYI @cmtice )

Does "multiple cus" in .o/dwo actually happen? 
From spec perspective I don't understand how this will work. Skeleton CU points to .o/dwo which then has multiple cus? What does it mean?

dwblaikie replied:
Ah, indeed - no, currently there's no well defined way to do LTO (LTO is basically the only time you get multiple CUs in a single .o straight out of the compiler) and Split DWARF (there's that long-standing issue that Cary and I should discuss how that should work & propose something to the DWARF committee, but we haven't as-yet).

I guess even if we did make that work, all the skeleton CUs in the dwo file would have the same DW_AT_dwo_name

dwblaikie commented on lldb/source/Plugins/SymbolFile/DWARF/SymbolFileDWARF.cpp:
```diff
@@ -1726,44 +1725,59 @@ lldb::ModuleSP SymbolFileDWARF::GetExternalModule(ConstString name) {
+  // This method can be called without going through the symbol vendor so we
+  // need to lock the module.
+  std::lock_guard<std::recursive_mutex> guard(GetModuleMutex());
+  SymbolFileDWARF *symbol_file = GetDIERefSymbolFile(die_ref);
+  if (symbol_file)
```
Similar feedback others have given elsewhere, roll the variable declaration into the `if` condition

dwblaikie commented on lldb/test/Shell/SymbolFile/DWARF/x86/dwp-foreign-type-units.cpp:
```diff
@@ -0,0 +1,91 @@
+
+// This test will make a type that will be compiled differently into two
+// different .dwo files in a type unit with the same type hash, but with
+// differing contents. I have discovered that the hash for the type unit is
+// simply based off of the typename and doesn't seem to differ when the contents
+// differ, so that will help us test foreign type units in the .debug_names
```
Rather than the personal statement, could maybe just be explicit:

"Clang's type unit signature is based only on the mangled name of the type, regardless of the contents of the type"

(I can tell you that's how it works - that's how I implemented it - it is a violation of the DWARF spec (the spec says to use a content hash - though that content hash is of the semantic contents, not the syntactic content (eg: if the TU contained a type with a single int member, the spec-compliant hash would be the same whether it was the type DIE followed by the int DIE, or the other way around) - so it would still be the same despite some variations in layout, so the index entries still wouldn't be portable to all matched-hash definitions))

dwblaikie commented on lldb/test/Shell/SymbolFile/DWARF/x86/dwp-foreign-type-units.cpp:
```diff
@@ -0,0 +1,91 @@
+// copy of the type unit ends up in the final .dwp file. To do this, LLDB will
+// look at the type unit and take the DWO name attribute and make sure it
+// matches, and if it doesn't, it will ignore the accelerator table entry.
+struct CustomType {
+  // We switch the order of "FloatType" and "IntegerType" so that if we do
+  // end up reading the wrong accelerator table entry, that we would end up
+  // getting an invalid offset and not find anything, or the offset would have
+  // matched and we would find the wrong thing.
```
Couldn't we have a difference between these type variants that's simpler - like one version of the struct with a single `int` member, and one version of the type with no members? Then do lookup of the `CustomType` and make sure the type we get back is the one with the `int` member? (or not, depending on the test)

ayermolo commented:
@clayborg 
Figured I reply here to your comment:
https://github.com/llvm/llvm-project/pull/88092#issuecomment-2059961175
This was regarding merging .debug_names in linker (although types are not implemented yet there), but FYI BOLT output is similar.
All the CUs are in one module.
Entry for type unit will have two indexes.
One for CU to which it belongs and another for type unit. Which can either be local or foreign.
example:
```
  Compilation Unit offsets [
    CU[0]: 0x00000000
    CU[1]: 0x00000029
  ]
  Foreign Type Unit signatures [
    ForeignTU[0]: 0x49dc260088be7e56
    ForeignTU[1]: 0x104ec427d2ebea6f
    ForeignTU[2]: 0xca1e65a66d92b970
    ForeignTU[3]: 0x104ec427d2ebea6f
  ]
  
  ...
  Bucket 1 [
    Name 1 {
      Hash: 0x7C96E4DB
      String: 0x00000027 "Foo2"
      Entry @ 0x117 {
        Abbrev: 0x1
        Tag: DW_TAG_structure_type
        DW_IDX_type_unit: 0x00
        DW_IDX_compile_unit: 0x00
        DW_IDX_die_offset: 0x00000021
        DW_IDX_parent: <parent not indexed>
      }
    }
```

clayborg committed:
Add support for BOLT generated .debug_names.

BOLT creates a single .debug_names table where foreign type units have both a DW_IDX_type_unit and a DW_IDX_compile_unit to uniquely identify the .dwo file that the type unit came from.

clayborg committed:
Finish merges with upstream HEAD.

clayborg committed:
Modified IsForeignTypeUnit to return a std::optional and modify to work with new DWARF changes.

clayborg committed:
Add tests for .dwo files.

clayborg committed:
After merge, address review comments and remove dead code.

clayborg committed:
Ran clang-format.

clayborg commented:
All feedback has been addressed and this now supports loading accelerator table entries from .dwo files for foreign type units as well.

ayermolo commented on llvm/lib/DebugInfo/DWARF/DWARFAcceleratorTable.cpp:
```diff
@@ -657,6 +657,42 @@ std::optional<uint64_t> DWARFDebugNames::Entry::getLocalTUOffset() const {
+std::optional<uint64_t>
+DWARFDebugNames::Entry::getForeignTUSkeletonCUOffset() const {
+  // Must have a DW_IDX_type_unit and it must be a foreign type unit.
+  if (!getForeignTUTypeSignature())
```
I think according to spec if there is only one TU DW_IDX_type_unit is not needed. Just like if there is only one CU.

clayborg replied:
How can you end up with a .debug_names table that contains only one type unit? If this can happen, I can modify getForeignTUTypeSignature() to do the right thing. But file would need to have a single foreign or local TU only, and no CUs at all?

clayborg replied:
Spec states:
```
6.1.1.4.2 List of CUs

The list of CUs immediately follows the header. Each entry in the list is an offset of the corresponding compilation unit in the .debug_info section. In the DWARF-32 format, a section offset is 4 bytes, while in the DWARF-64 format, a section offset is 8 bytes.

The total number of entries in the list is given by comp_unit_count. There must be at least one CU.
```

clayborg replied:
It is still necessary otherwise we will think it is just the one and only `DW_IDX_comp_unit`, not the one any only `DW_IDX_type_unit`...

labath commented on lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp:
```diff
@@ -48,20 +60,84 @@ DebugNamesDWARFIndex::GetUnits(const DebugNames &debug_names) {
+    // file from it and get the type unit by signature from there. If we find
+    // the type unit in the .dwo file, we don't need to check that the
+    // DW_AT_dwo_name matches because each .dwo file can have its own type unit.
+    std::optional<uint64_t> unit_offset = entry.getForeignTUSkeletonCUOffset();
+    if (!unit_offset)
+      return nullptr; // Return NULL, this is a type unit, but couldn't find it.
+    DWARFUnit *cu =
+        m_debug_info.GetUnitAtOffset(DIERef::Section::DebugInfo, *unit_offset);
+    if (!cu)
+      return nullptr; // Return NULL, this is a type unit, but couldn't find it.
```
I think this part could be moved into the common (dwp and non-dwp) code.

labath commented on lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.cpp:
```diff
@@ -48,20 +60,84 @@ DebugNamesDWARFIndex::GetUnits(const DebugNames &debug_names) {
+  if (!foreign_tu)
+    return nullptr; // Return NULL, this is a type unit, but couldn't find it.
+
+  std::optional<uint64_t> cu_offset = entry.getForeignTUSkeletonCUOffset();
+  if (cu_offset) {
+    DWARFUnit *cu = m_debug_info.GetUnitAtOffset(DIERef::DebugInfo, *cu_offset);
+    if (cu) {
```
and merged with this.

labath commented on lldb/source/Plugins/SymbolFile/DWARF/DebugNamesDWARFIndex.h:
```diff
@@ -85,6 +86,31 @@ class DebugNamesDWARFIndex : public DWARFIndex {
+  ///     doesn't match the originating skeleton compile unit's entry
+  ///   Returns std::nullopt if this entry is not a foreign type unit entry.
+  std::optional<DWARFTypeUnit *>
+  IsForeignTypeUnit(const DebugNames::Entry &entry) const;
```
Nit: I would expect a function called `IsSomething` to return `bool`. Maybe just call it `GetForeignTypeUnit` ?

labath commented on lldb/source/Plugins/SymbolFile/DWARF/SymbolFileDWARF.cpp:
```diff
@@ -1741,45 +1741,61 @@ lldb::ModuleSP SymbolFileDWARF::GetExternalModule(ConstString name) {
+
+  // If we are currently in a .dwo file and our file index doesn't match we need
+  // to let the base symbol file handle this.
+  SymbolFileDWARFDwo *dwo = llvm::dyn_cast_or_null<SymbolFileDWARFDwo>(this);
+  if (dwo)
+    return dwo->GetBaseSymbolFile().GetDIERefSymbolFile(die_ref);
```
A better (and more consistent) way to implement this would be to make `GetDIERefSymbolFile` virtual, and then have the Dwo implementation do `return GetBaseSymbolFile().GetDIERefSymbolFile(die_ref)`.

paperchalice committed:
Revert "[Misc] Use `LLVM_ENABLE_ABI_BREAKING_CHECKS` correctly" (#94982)

Reverts llvm/llvm-project#94212
Some codes assume that `NDEBUG` implies
`LLVM_ENABLE_ABI_BREAKING_CHECKS`, thus #94212 breaks some build bots.

clayborg committed:
Respond to user comments.

clayborg committed:
Address user feedback.

- Make GetDIERefSymbolFile() virtual in SymbolFileDWARF and override in SymbolFileDWARFDwo
- Rename IsForeignTypeUnit to GetForeignTypeUnit

clayborg commented:
I think I have addressed all issues. Does anyone have time to review?

labath reviewed (commented):
Most of the patch is very clean, but I'm bothered by the `getForeignTUSkeletonCUOffset` function, how it opens up with a mostly redundant (the callers checks this already) call to `getForeignTUTypeSignature`, and then proceeds with a reimplementation of `getCUOffset` (sans the type unit check). I think we could design a better set of APIs for this, but I'm not sure what those is, because I'm unclear of the meaning of various combinations of DW_IDX unit entries.

What the new function (I think) essentially does is "give me the CU associated with this entry even if the entry does not describe a DIE in this CU" (because `DW_IDX_die_offset` is relative to a type unit)". I think this API would make sense, even without needing to talk about type units. However, is it actually implementable? For single-CU indexes, that's fine, because we can kind of assume all entries are related to that CU. But what about multi-CU indexes? The only way to determine the CU there is to have an explicit attribute. Are we saying that each entry in a multi-CU index must have a DW_IDX_compile_unit (in addition to a DW_IDX_type_unit)?

If the answer is yes, then this function can be implemented, but then I think the current implementation of `getCUOffset` (which I interpret as "give me the CU offset **IF** this entry is relative to that CU") doesn't make sense -- because it treats entries with an explicit `DW_IDX_compile_unit` different from an implicit/missing `DW_IDX_compile_unit`. And in this world, we assume that `DW_IDX_type_unit` takes precedence over `DW_IDX_compile_unit` -- if both are present then `DW_IDX_die_offset` is relative to the former. And I'm not sure if we would actually want to take up space by putting both values into the index. Looking at existing producers would be interesting, but I'm not sure if there are any (lld does not merge type unit indexes right now, possibly for this very reason). Maybe one could produce something with LTO?

OTOH, if the answer is no, then the function isn't implementable in the generic case. That doesn't mean you can't implement this feature -- which is useful for guarding against ODR violations (exacerbated by llvm's nonconforming type signature computation algorithm...). However, I think it could be implemented in a much simpler way. For example, lldb could just check if it's looking at a single-CU index, and get the CU offset that way. No extra llvm APIs would be needed.

ayermolo commented:
> Most of the patch is very clean, but I'm bothered by the `getForeignTUSkeletonCUOffset` function, how it opens up with a mostly redundant (the callers checks this already) call to `getForeignTUTypeSignature`, and then proceeds with a reimplementation of `getCUOffset` (sans the type unit check). I think we could design a better set of APIs for this, but I'm not sure what those is, because I'm unclear of the meaning of various combinations of DW_IDX unit entries.
> 
> What the new function (I think) essentially does is "give me the CU associated with this entry even if the entry does not describe a DIE in this CU" (because `DW_IDX_die_offset` is relative to a type unit)". I think this API would make sense, even without needing to talk about type units. However, is it actually implementable? For single-CU indexes, that's fine, because we can kind of assume all entries are related to that CU. But what about multi-CU indexes? The only way to determine the CU there is to have an explicit attribute. Are we saying that each entry in a multi-CU index must have a DW_IDX_compile_unit (in addition to a DW_IDX_type_unit)?
> 
> If the answer is yes, then this function can be implemented, but then I think the current implementation of `getCUOffset` (which I interpret as "give me the CU offset **IF** this entry is relative to that CU") doesn't make sense -- because it treats entries with an explicit `DW_IDX_compile_unit` different from an implicit/missing `DW_IDX_compile_unit`. And in this world, we assume that `DW_IDX_type_unit` takes precedence over `DW_IDX_compile_unit` -- if both are present then `DW_IDX_die_offset` is relative to the former. And I'm not sure if we would actually want to take up space by putting both values into the index. Looking at existing producers would be interesting, but I'm not sure if there are any (lld does not merge type unit indexes right now, possibly for this very reason). Maybe one could produce something with LTO?
> 
> OTOH, if the answer is no, then the function isn't implementable in the generic case. That doesn't mean you can't implement this feature -- which is useful for guarding against ODR violations (exacerbated by llvm's nonconforming type signature computation algorithm...). However, I think it could be implemented in a much simpler way. For example, lldb could just check if it's looking at a single-CU index, and get the CU offset that way. No extra llvm APIs would be needed.

BOLT does as I mentioned in:
https://github.com/llvm/llvm-project/pull/87740#issuecomment-2060023287

labath commented:
> BOLT does as I mentioned in: [#87740 (comment)](https://github.com/llvm/llvm-project/pull/87740#issuecomment-2060023287)

I'm sorry, I missed that part. I'm bit of late to the party here.

> 
> It will put all the CUs/Tus into one one module, thus for foreign TUs both DW_IDX_compile_unit and DW_IDX_type_unit is needed.

Right, the CU is necessary to find the DWO file containing the type unit. I forgot about that. That resolves _some_ of my questions. However, I still find the current implementation of `getCUOffset` strange, namely that:
- if the entry contains a `DW_IDX_compile_unit` (perhaps because we have a multi-cu index), it returns that CU, regardless of anything else
- if it does not contain the attribute (and we have a single-CU index), it returns that CU *only* if the index also does not contain a DW_IDX_type_unit (even though the standard says "In a per-CU index, index entries without this attribute implicitly refer to the single CU")

This doesn't seem particularly useful. I think it should either always return CU offset (regardless of whether it's implicit or explicit) or *only* return it if the DW_IDX_type_unit is not present. And of these two, I think the first one is better, because the caller can always explicitly check for the presence of the type attribute, and it avoids the need to introduce another function (which is needed because we now have a caller who wants the CU offset even though the entry refers to a type unit).

From what I can see, that type unit check was added relatively recently (#72952), and it was to support this piece of code in lldb:
```
+  // Look for a DWARF unit offset (CU offset or local TU offset) as they are
+  // both offsets into the .debug_info section.
+  std::optional<uint64_t> unit_offset = entry.getCUOffset();
+  if (!unit_offset) {
+    unit_offset = entry.getLocalTUOffset();
+    if (!unit_offset)
+      return std::nullopt;
+  }
```
If we just reversed the order of those checks in lldb, then it wouldn't matter that this function return a CU offset even if the index entry does not refer to a DIE in that CU.

Or am I (still) missing something?

clayborg commented:
> Most of the patch is very clean, but I'm bothered by the `getForeignTUSkeletonCUOffset` function, how it opens up with a mostly redundant (the callers checks this already) call to `getForeignTUTypeSignature`, and then proceeds with a reimplementation of `getCUOffset` (sans the type unit check). I think we could design a better set of APIs for this, but I'm not sure what those is, because I'm unclear of the meaning of various combinations of DW_IDX unit entries.

If we have an entry with a `DW_IDX_type_unit`, it can be a normal type unit, or a foreign type unit, that depends on the indexes value. If it is less than the number of local type units, then it is a local type unit and there should be no `DW_IDX_comp_unit` in the entry, else it is a foreign type unit from a .dwo or .dwp file. If we have a .dwp file, we need the `DW_IDX_comp_unit` to be able to tell if the foreign type unit made it into the .dwp file, or if this is an entry that represents the a different type unit and we should ignore it.  If we have no .dwp file, then all entries for foreign type units are valid and we don't need to check the compile unit it came from. 

So if you just ask an entry for its `DW_IDX_comp_unit`, you will get an answer for an entry from a normal compile unit, or for a foreign type unit with a compile unit that will be used to discriminate for the .dwp file case. 

So this function is named `getForeignTUSkeletonCUOffset()` to help say "this might have  compile unit entry, but we only want it if this is a foreign type unit. I took me a while to understand the spec when it came to foreign type units...

> What the new function (I think) essentially does is "give me the CU associated with this entry even if the entry does not describe a DIE in this CU" (because `DW_IDX_die_offset` is relative to a type unit)". I think this API would make sense, even without needing to talk about type units. However, is it actually implementable? For single-CU indexes, that's fine, because we can kind of assume all entries are related to that CU. But what about multi-CU indexes? The only way to determine the CU there is to have an explicit attribute. Are we saying that each entry in a multi-CU index must have a DW_IDX_compile_unit (in addition to a DW_IDX_type_unit)?

Yes, but only if the `DW_IDX_type_unit` represents a foreign type unit, and we will only need the `DW_IDX_comp_unit` if we have a .dwp file. But since we don't know if the user will use a .dwp file, we always need to put the data for the originating compile uint in as a `DW_IDX_comp_unit`.

I would rather have had a `DW_IDX_skeleton_unit` to differentiate between a `DW_IDX_comp_unit` and requiring the user to know that they can't just find the compile unit, but they must find the non-skeleton compile unit,  before adding the `DW_IDX_die_offset`. But we don't have that in the spec...

> 
> If the answer is yes, then this function can be implemented, but then I think the current implementation of `getCUOffset` (which I interpret as "give me the CU offset **IF** this entry is relative to that CU") doesn't make sense -- because it treats entries with an explicit `DW_IDX_compile_unit` different from an implicit/missing `DW_IDX_compile_unit`. 

It doesn't treat them differently. if you ask an entry for its `DW_IDX_compile_unit` and it doesn't have one, we can fall back to grabbing a CU from the CU table, but only if the CU table has only 1 entry in it.

> And in this world, we assume that `DW_IDX_type_unit` takes precedence over `DW_IDX_compile_unit` -- if both are present then `DW_IDX_die_offset` is relative to the former. And I'm not sure if we would actually want to take up space by putting both values into the index. Looking at existing producers would be interesting, but I'm not sure if there are any (lld does not merge type unit indexes right now, possibly for this very reason). Maybe one could produce something with LTO?

LTO and bolt and soon lld will produce these single tables with multiple compile units. 

And we MUST have the `DW_IDX_comp_unit` for foreign type units because type units can generate differing content and one copy of the type unit will make it into the .dwp and we need to know from which .dwo file it came and only use the entries from the matching .dwo file because the offsets can be meaningless for type units with the same signature but that came from different .dwo files.

> 
> OTOH, if the answer is no, then the function isn't implementable in the generic case. That doesn't mean you can't implement this feature -- which is useful for guarding against ODR violations (exacerbated by llvm's nonconforming type signature computation algorithm...). However, I think it could be implemented in a much simpler way. For example, lldb could just check if it's looking at a single-CU index, and get the CU offset that way. No extra llvm APIs would be needed.

All of these APIs get the single CU if there is only one for both the case where you only have a `DW_IDX_comp_unit` in an entry and for the case where you have a `DW_IDX_type_unit` and/or not a `DW_IDX_comp_unit` in the same entry.

So not sure what you are asking for here, but this is the best way I can represent the messy state of the .debug_names tables. Monolithic tables with multiple CUs are here now from BOLT, and will soon be from lld, so we need to be able to handle these.

labath commented:
My previous message reflected my confusion about the usefulness about the DW_IDX_cu+DW_IDX_tu combo. Let me try to start over.

The thing I don't like about this patch is the duplication between `getForeignTUSkeletonCUOffset` (the new function) and `getCUOffset` (the existing function). Since the duplication is motivated by undesirable behavior (result) of the existing function in some cases, this got me thinking if there's a way to define a single function in a way that will be usable in all scenarios (and still have a consistent easy-to-understand definition).

To see if that's possible, let's consider the possible cases (I'm putting cases with DW_IDX_tu first, as those are the interesting ones here):

1. A single-CU index, DW_IDX_cu not present, DW_IDX_tu present
2. A multi-CU index, DW_IDX_cu and DW_IDX_tu present

These two definitely are the most common (of those including type units) ones, but we can also have:

3. A single-CU index, DW_IDX_cu and DW_IDX_tu present
4. A multi-CU index, DW_IDX_cu not present, DW_IDX_tu present

For completeness, we can also have the cases without type units. I will include only include one here, as I think the others are obvious:

5. A single-CU index, DW_IDX_cu and DW_IDX_tu **not** present

Now that we have that, what are the possible `getCUOffset`? I can think of three:

a) "return the value of DW_IDX_cu. Period." If an explicit value is not present return nothing. I don't think anyone wants this, but I'm mentioning it for completeness. With this definition, the function should return the cu offset only in the second and third cases (`NYYNN`)
b) "return the DW_IDX_cu value, or the single CU from the index." Let's call this the "raw" version. With this definition, I think this function should return the cu offset in all cases except the fourth (that one is impossible to resolve): `YYYNY`
c) "return the DW_IDX_cu value, only if the entry describes a DIE within that CU." Let's call this the "semantic" version. I believe a function with this definition should not return a CU offset only in the last case: `NNNNY`

Now, the problem I see is that the current implementation of `getCUOffset` is **neither** of these options. If an entry has an (explicit) DW_IDX_cu, it will always return it (even if it also contains DW_IDX_tu). However, an implicit CU will only be returned if DW_IDX_tu is not present. I.e, this function does a `NYNNY`.

Therefore, my proposal is to redefine the function to do (b), which essentially amounts to deleting these two lines:
```
  if (lookup(dwarf::DW_IDX_type_unit).has_value())
    return std::nullopt;
```

Besides providing a consistent definition of the function (YMMV, if you can provide an explanation for the current behavior, I'd like to hear it), it would also enable you to reuse it in `getForeignTUSkeletonCUOffset`, as that is exactly the behavior you need there (although, at that point, I don't think `getForeignTUSkeletonCUOffset` needs to exist as a separate function, as it will be simple enough to inline into lldb).

Changing the definition of course requires going through the call sites and making sure they are okay with that definition. I counted a total of three, one in llvm and two in lldb. The call in llvm (llvm-dwarfdump.cpp) is broken for type units, both before and after this change. Maybe the same goes for one of the calls in lldb (DebugNamesDWARFIndex::GetGlobalVariables -- I'm not sure if you can have global (static class member) variables in a type unit). The last one is definitely broken (and the test suite attests that, but fixing it is trivial: see https://github.com/llvm/llvm-project/commit/c5c2a9d61132c81756267694dd7af52d0a1121d9. I'll also note that this code was also broken for the case (2) -- a multi CU-index with the accelerator entry referring to a DIE in a (local) type unit, and that that patch fixes that.

cmtice commented:
Just a reminder: Once we complete the work of adding type units to the merged debug_names index, foreign type unit entries will need (and have) DW_IDX_compile_unit values, even in the case without .dwp files, and will need to be handled properly. So that should be kept in mind while implementing this.

clayborg commented:
> My previous message reflected my confusion about the usefulness about the DW_IDX_cu+DW_IDX_tu combo. Let me try to start over.
> 
> The thing I don't like about this patch is the duplication between `getForeignTUSkeletonCUOffset` (the new function) and `getCUOffset` (the existing function). Since the duplication is motivated by undesirable behavior (result) of the existing function in some cases, this got me thinking if there's a way to define a single function in a way that will be usable in all scenarios (and still have a consistent easy-to-understand definition).
> 
> To see if that's possible, let's consider the possible cases (I'm putting cases with DW_IDX_tu first, as those are the interesting ones here):
> 
> 1. A single-CU index, DW_IDX_cu not present, DW_IDX_tu present
> 2. A multi-CU index, DW_IDX_cu and DW_IDX_tu present
> 
> These two definitely are the most common (of those including type units) ones, but we can also have:
> 
> 3. A single-CU index, DW_IDX_cu and DW_IDX_tu present
> 4. A multi-CU index, DW_IDX_cu not present, DW_IDX_tu present
> 
> For completeness, we can also have the cases without type units. I will include only include one here, as I think the others are obvious:
> 
> 5. A single-CU index, DW_IDX_cu and DW_IDX_tu **not** present
> 
> Now that we have that, what are the possible `getCUOffset`? I can think of three:
> 
> a) "return the value of DW_IDX_cu. Period." If an explicit value is not present return nothing. I don't think anyone wants this, but I'm mentioning it for completeness. With this definition, the function should return the cu offset only in the second and third cases (`NYYNN`) 
> b) "return the DW_IDX_cu value, or the single CU from the index." Let's call this the "raw" version. With this definition, I think this function should return the cu offset in all cases except the fourth (that one is impossible to resolve): `YYYNY` 
> c) "return the DW_IDX_cu value, only if the entry describes a DIE within that CU." Let's call this the "semantic" version. I believe a function with this definition should not return a CU offset only in the last case: `NNNNY`
> 
> Now, the problem I see is that the current implementation of `getCUOffset` is **neither** of these options. If an entry has an (explicit) DW_IDX_cu, it will always return it (even if it also contains DW_IDX_tu). However, an implicit CU will only be returned if DW_IDX_tu is not present. I.e, this function does a `NYNNY`.
> 
> Therefore, my proposal is to redefine the function to do (b), which essentially amounts to deleting these two lines:
> 
> ```
>   if (lookup(dwarf::DW_IDX_type_unit).has_value())
>     return std::nullopt;
> ```

But we won't want a CU index back if we have a `DW_IDX_type_unit` that is a local type unit. Makes no sense to return a CUOffset from such entries. So maybe we modify this be:

```
std::optional<uint64_t> DWARFDebugNames::Entry::getCUIndex() const {
  if (std::optional<DWARFFormValue> Off = lookup(dwarf::DW_IDX_compile_unit))
    return Off->getAsUnsignedConstant();
  // In a per-CU index, the entries without a DW_IDX_compile_unit attribute
  // implicitly refer to the single CU, but only if we don't have a
  // DW_IDX_type_unit that is a foreign type unit.
  if (IsForeignTypeUnit())
    return std::nullopt;
  if (NameIdx->getCUCount() == 1)
    return 0;
  return std::nullopt;
}
```
Then we implement a `bool IsForeignTypeUnit() const` that just checks if there is a DW_IDX_type_unit whose index is a foreign type unit.

> 
> Besides providing a consistent definition of the function (YMMV, if you can provide an explanation for the current behavior, I'd like to hear it), it would also enable you to reuse it in `getForeignTUSkeletonCUOffset`, as that is exactly the behavior you need there (although, at that point, I don't think `getForeignTUSkeletonCUOffset` needs to exist as a separate function, as it will be simple enough to inline into lldb).

That is fine with me as long as you agree with the above fix.

Let me know if the above fix to `DWARFDebugNames::Entry::getCUIndex()` makes sense.

labath commented:
> But we won't want a CU index back if we have a `DW_IDX_type_unit` that is a local type unit. Makes no sense to return a CUOffset from such entries.

Can you elaborate on that? I believe it doesn't make sense to you, but I just don't see the logic behind it. In particular I don't understand why it makes sense to return the CU offset when the DW_IDX_compile_unit attribute is explicitly specified (cases 2 and 3), but not when the compile unit is referred to implicitly via the lack of the attribute on a single-cu index (case 1)

clayborg commented:
> > But we won't want a CU index back if we have a `DW_IDX_type_unit` that is a local type unit. Makes no sense to return a CUOffset from such entries.
> 
> Can you elaborate on that? I believe it doesn't make sense to you, but I just don't see the logic behind it. In particular I don't understand why it makes sense to return the CU offset when the DW_IDX_compile_unit attribute is explicitly specified (cases 2 and 3), but not when the compile unit is referred to implicitly via the lack of the attribute on a single-cu index (case 1)

When we have an entry like:
```
DW_IDX_type_unit(0)
DW_IDX_die_offset(0x32)
```
And type unit at index 0 is a local type unit, then the entry describes a type unit DIE in the .debug_info of the current executable. The current API we have is in `DWARFAcceleratorTable.h` for `DWARFDebugNames::Entry` is:
```
    std::optional<uint64_t> getCUOffset() const override;
    std::optional<uint64_t> getLocalTUOffset() const override;
    std::optional<uint64_t> getForeignTUTypeSignature() const override;
```

So if we have an entry mentioned above where the type unit lives in the local .debug_info section, we should get an invalid value back from a call to `getCUOffset()` if this is a local type unit. And we should get a valid value back from `getLocalTUOffset()`. But it makes no sense to return the first compile unit for a call to `getCUOffset()` for an entry that represents a local type unit because it isn't related to the local type unit. 

Lets say we do allow an entry with a local type unit to return the first CU from the .debug_names only because there is only one CU in the .debug_names table. Soon `lld` will generate a single .debug_names table with multiple CUs. We will get no CU offset back for an entry that contains a local type unit + die offset (no CU specified) when using a combined .debug_names table because it can't rely on there only being one CU in the list, but we will get a valid value back for individual .debug_names tables, which makes the API usage inconsistent between individual tables and combined tables.

Any foreign type units that must be associated with a skeleton compile unit, should have both a `DW_IDX_type_unit` and a `DW_IDX_comp_unit` to make sure we can make this connection between a TU and originating CU.

labath commented:
Thank you. I now have a better idea of where you're coming from.

> > Can you elaborate on that? I believe it doesn't make sense to you, but I just don't see the logic behind it. In particular I don't understand why it makes sense to return the CU offset when the DW_IDX_compile_unit attribute is explicitly specified (cases 2 and 3), but not when the compile unit is referred to implicitly via the lack of the attribute on a single-cu index (case 1)
> 
> When we have an entry like:
> 
> ```
> DW_IDX_type_unit(0)
> DW_IDX_die_offset(0x32)
> ```
> 
> And type unit at index 0 is a local type unit, then the entry describes a type unit DIE in the .debug_info of the current executable. The current API we have is in `DWARFAcceleratorTable.h` for `DWARFDebugNames::Entry` is:
> 
> ```
>     std::optional<uint64_t> getCUOffset() const override;
>     std::optional<uint64_t> getLocalTUOffset() const override;
>     std::optional<uint64_t> getForeignTUTypeSignature() const override;
> ```
> 
> So if we have an entry mentioned above where the type unit lives in the local .debug_info section, we should get an invalid value back from a call to `getCUOffset()` if this is a local type unit. And we should get a valid value back from `getLocalTUOffset()`. But it makes no sense to return the first compile unit for a call to `getCUOffset()` for an entry that represents a local type unit because it isn't related to the local type unit.

This sounds to me like you'd prefer that getCUOffset behave like the (c) definition from my post above: "return the DW_IDX_cu value, only if the entry describes a DIE within that CU". Do you concur?

If so I think mutually acceptable version (the thing I want to avoid the most is the duplicate implementation of DW_IDX_cu retrieval) could be to have flavours of getCUOffset, a higher-level version doing (c), and a lower-level version doing (b), which would be called into from (c). Something like:
```
// Returns the CU that is related to this entry, not necessarily the CU that *contains* the entry.
getRelatedCUIndex /*name TBD*/() {
  if (std::optional<DWARFFormValue> Off = lookup(dwarf::DW_IDX_compile_unit))
    return Off->getAsUnsignedConstant();
  if (NameIdx->getCUCount() == 1)
    return 0;
  return std::nullopt;
}
// ditto
getRelatedCUOffset() {
  std::optional<uint64_t> Index = getCUIndex();
  if (!Index || *Index >= NameIdx->getCUCount())
    return std::nullopt;
  return NameIdx->getCUOffset(*Index);
}
// Returns the CU offset, if this entry describes a DIE in that CU
getCUOffset() {
  if (lookup(dwarf::DW_IDX_type_unit).has_value())
    return std::nullopt;
  // Entries without a type unit attribute describe a DIE in the compile unit.
  return getRelatedCUOffset();
}
```

and `getForeignTUSkeletonCUOffset` could be:
```
  if (!getForeignTUTypeSignature()) return std::nullopt;
  return getRelatedCUOffset();
```
... although I'd prefer to not have that function on the llvm interface, but rather inline it into lldb.

How does that sound?

> 
> Lets say we do allow an entry with a local type unit to return the first CU from the .debug_names only because there is only one CU in the .debug_names table. Soon `lld` will generate a single .debug_names table with multiple CUs. We will get no CU offset back for an entry that contains a local type unit + die offset (no CU specified) when using a combined .debug_names table because it can't rely on there only being one CU in the list, but we will get a valid value back for individual .debug_names tables, which makes the API usage inconsistent between individual tables and combined tables.

Okay, I can see how that can appear to be inconsistent. However, at the same time, I think it's the most accurate representation of what's in the debug info. If you have a type unit in a multi-CU index without a DW_IDX_cu attribute, then you really cannot tell which CU does it relate to. It doesn't matter if it a local or foreign type unit -- the information is just not there. You may not *need* this information for local type units, but I don't think that means we should deliberately suppress it -- particularly if it means that the function can be used as building block to implement higher level functionality (like I did with `getRelatedCUOffset` above)

> 
> Any foreign type units that must be associated with a skeleton compile unit, should have both a `DW_IDX_type_unit` and a `DW_IDX_comp_unit` to make sure we can make this connection between a TU and originating CU.

I'm not sure if this refers to multi-CU indexes or indexes in general, but if it's the former, then I totally agree. (In a single-CU index, I think that connection can be given implicitly through the sole CU in the index)

clayborg commented:
Got messed up trying to sync with upstream. I will abandon this and create a new PR that references this one...

ayermolo commented:
Probably doing
git fetch upstream
git rebase upstream/main
git push -f 
will fix it (assuming your branch is tracking your repo)

clayborg committed:
Address user feedback.

clayborg commented:
I messed up the branch on this PR. New PR with the requested fixes from @labath at https://github.com/llvm/llvm-project/pull/96243

clayborg commented:
@ayermolo: you suggestion worked. thanks. This PR is back on track...

@labath: I implemented your suggestions, let me know if this look ok now

labath reviewed (commented):
The llvm stuff looks good (thank you), but there are still (small) some unresolved comments in the lldb code. I've pinged them in this review.

clayborg commented on lldb/source/Plugins/SymbolFile/DWARF/SymbolFileDWARF.cpp:
```diff
@@ -1727,45 +1727,61 @@ lldb::ModuleSP SymbolFileDWARF::GetExternalModule(ConstString name) {
+  // This method can be called without going through the symbol vendor so we
+  // need to lock the module.
+  std::lock_guard<std::recursive_mutex> guard(GetModuleMutex());
+  SymbolFileDWARF *symbol_file = GetDIERefSymbolFile(die_ref);
```
It only uses the file index component, but that component might not be set in the `DIERef` object and if it isn't, it returns the current SymbolFileDWARF object. We can make a function like:
```
SymbolFileDWARF *GetSymbolFileByFileIndex(std::optional<uint32_t> file_index);
```
The file index only makes sense if it comes from a DIERef object as it understands the value and knows the magic value for the .dwp file, so the above API is much less clear as who else knows what a file index is and would actaully know to ask for the symbol file using such an index? Seems less clear to me than the existing API. Let me know if feel we should still change this

labath replied:
I don't think the file index needs to necessarily come from a DIERef. I can imagine other scenarios where it might be interesting to ask for a symbol file, but I think we can deal with it when the time comes.

clayborg committed:
Respond to user feeback.

clayborg commented:
All issues resolved except the change of `GetDIERefSymbolFile`. Let me know if you agree with my comments, or want a `GetSymbolFileByFileIndex(std::optional<uint32_t> file_idx)`.

labath commented on lldb/source/Plugins/SymbolFile/DWARF/SymbolFileDWARF.h:
```diff
@@ -241,6 +241,15 @@ class SymbolFileDWARF : public SymbolFileCommon {
+  /// offset makes sense in the DIERef.
+  virtual SymbolFileDWARF *GetDIERefSymbolFile(const DIERef &die_ref);
+
   virtual DWARFDIE GetDIE(const DIERef &die_ref);
```
Now GetDIE is virtual and handles SymbolFile resolution, this one doesn't need to be.

clayborg commented:
An extra character snuck in and messed with the buildbots, fixed with:
```
commit fc066ca1c32b4aef549f3e371dc70589804aba0f (HEAD -> main, origin/main, origin/HEAD)
Author: Greg Clayton <clayborg@gmail.com>
Date:   Mon Jun 24 10:15:55 2024 -0700

    Fix buildbots for https://github.com/llvm/llvm-project/pull/87740
```

labath commented:
It looks like the test is flaky:     https://lab.llvm.org/buildbot/#/builders/59/builds/535

It looks like the order of the types is nondeterministic. I think you should be able to use CHECK-DAG along with [this trick to match newlines](https://llvm.org/docs/CommandGuide/FileCheck.html#matching-newline-characters) to make the test accept both orderings.

vvereschaka commented:
This build got the same failed test on the windows host - https://lab.llvm.org/buildbot/#/builders/141/builds/318

```
Failed Tests (1):
  lldb-shell :: SymbolFile/DWARF/x86/dwp-foreign-type-units.cpp
```

```
# .---command stderr------------
# | C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\llvm-project\lldb\test\Shell\SymbolFile\DWARF\x86\dwp-foreign-type-units.cpp:34:16: error: NODWP-NEXT: is not on the line after the previous match
# | // NODWP-NEXT: unsigned int
# |                ^
# | <stdin>:20:10: note: 'next' match was here
# |  typedef unsigned int IntegerType;
# |          ^
# | <stdin>:7:13: note: previous match ended here
# | unsigned int
# |             ^
# | <stdin>:8:1: note: non-matching line after previous match is here
# | int
# | ^
# | 
# | Input file: <stdin>
# | Check file: C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\llvm-project\lldb\test\Shell\SymbolFile\DWARF\x86\dwp-foreign-type-units.cpp
# | 
# | -dump-input=help explains the following input dump.
# | 
# | Input was:
# | <<<<<<
# |          .
# |          .
# |          .
# |         15:  typedef double FloatType; 
# |         16:  CustomType::IntegerType x; 
# |         17:  CustomType::FloatType y; 
# |         18: } 
# |         19: struct CustomType { 
# |         20:  typedef unsigned int IntegerType; 
# | next:34              !~~~~~~~~~~~               error: match on wrong line
# |         21:  typedef float FloatType; 
# |         22:  CustomType::IntegerType x; 
# |         23:  CustomType::FloatType y; 
# |         24: } 
# | >>>>>>
# `-----------------------------
# error: command failed with exit status: 1
```

dwblaikie commented:
> It looks like the test is flaky: https://lab.llvm.org/buildbot/#/builders/59/builds/535
> 
> It looks like the order of the types is nondeterministic. I think you should be able to use CHECK-DAG along with [this trick to match newlines](https://llvm.org/docs/CommandGuide/FileCheck.html#matching-newline-characters) to make the test accept both orderings.

What are lldb's guarantees in this regard? Clang/LLVM/etc require nondeterministic output - presumably the same would be valuable to lldb, but I don't know what lldb's precedents are in this regard.

clayborg commented:
Here is a PR to fix the flaky test:

https://github.com/llvm/llvm-project/pull/96800

labath commented:
> > It looks like the test is flaky: https://lab.llvm.org/buildbot/#/builders/59/builds/535
> > It looks like the order of the types is nondeterministic. I think you should be able to use CHECK-DAG along with [this trick to match newlines](https://llvm.org/docs/CommandGuide/FileCheck.html#matching-newline-characters) to make the test accept both orderings.
> 
> What are lldb's guarantees in this regard? Clang/LLVM/etc require nondeterministic output - presumably the same would be valuable to lldb, but I don't know what lldb's precedents are in this regard.

It would be useful, but we've never made a serious effort to try to achieve that.
